Close support cover artwork in Now Playing module, #5336

low-batt opened the pull request:
This commit will:
- Move `NowPlayingInfoManager` out of `PlayerCore` and into its own file
- Change `PlayerCore.pauseChanged` to call `syncUITime` before changing the state so that `PlaybackInfo.videoPosition` is up to date when `NowPlayingInfoManager.updateInfo` is called
- Change updateInfo to set `MPNowPlayingInfoPropertyPlaybackRate` to zero when playback is paused as required by Now Playing
- Change updateInfo to set `playbackState` to `playing` before setting it to `paused` when playback is paused from the start to workaround a Now Playing problem
- Remove the state parameter from the `updateInfo` method and change the method to use the `state` property in `PlaybackInfo`
- Change `updateInfo` to remove `MPMediaItemPropertyAlbumTitle` and `MPMediaItemPropertyArtist` when not needed instead of setting them to the empty string
- Add a `cgImage` property and a `resized` method to the `NSImage` extension
- Add a `CGImage` extension
- Add a `CGSize` extension
- Add a `MPNowPlayingPlaybackState` extension that adopts the `CustomStringConvertible` property to allow values to be easily logged
- Add an `isImage` property to `MPVTrack`
- Add a `readArtwork` method to `FFmpegController`
- Add calls to `NowPlayingInfoManager.updateInfo` when thumbnails are ready
- Add support for setting `MPMediaItemPropertyArtwork` in `nowPlayingInfo` to `NowPlayingInfoManager`
- Add a new `enableNowPlayingArtwork` internal setting that can be used to disable showing artwork should a serious problem be discovered

This commit is builds upon the pioneering work done by @svobs to add support for showing artwork in Now Playing. NowPlayingInfoManager will first try and find a video track that represents cover art and use that for the Now Playing artwork. If that fails it will then ask the macOS Quick Look system to generate a thumbnail to use as Now Playing artwork. If that fails then it will try and use an OSC thumbnail.

This commit also corrects several existing problems with Now Playing found during testing.

- [x] I have read [CONTRIBUTING.md](https://github.com/iina/iina/blob/develop/CONTRIBUTING.md)
- [x] This implements/fixes issue #5336.

---

**Description:**

low-batt committed:
Close support cover artwork in Now Playing module, #5336

This commit will:
- Move NowPlayingInfoManager out of PlayerCore and into its own file
- Change PlayerCore.pauseChanged to call syncUITime before changing the
  state so that PlaybackInfo.videoPosition is up to date when
  NowPlayingInfoManager.updateInfo is called
- Change updateInfo to set MPNowPlayingInfoPropertyPlaybackRate to zero
  when playback is paused as required by Now Playing
- Change updateInfo to set playbackState to playing before setting it to
  paused when playback is paused from the start to workaround a
  Now Playing problem
- Remove the state parameter from the updateInfo method and change the
  method to use the state property in PlaybackInfo
- Change updateInfo to remove MPMediaItemPropertyAlbumTitle and
  MPMediaItemPropertyArtist when not needed instead of setting them to
  the empty string
- Add a cgImage property and a resized method to the NSImage extension
- Add a CGImage extension
- Add a CGSize extension
- Add a MPNowPlayingPlaybackState extension that adopts the
  CustomStringConvertible property to allow values to be easily logged
- Add an isImage property to MPVTrack
- Add a readArtwork method to FFmpegController
- Add calls to NowPlayingInfoManager.updateInfo when thumbnails are
  ready
- Add support for setting MPMediaItemPropertyArtwork in nowPlayingInfo
  to NowPlayingInfoManager
- Add a new enableNowPlayingArtwork internal setting that can be used to
  disable showing artwork should a serious problem be discovered

This commit is builds upon the pioneering work done by @svobs to add
support for showing artwork in Now Playing. NowPlayingInfoManager will
first try and find a video track that represents cover art and use that
for the Now Playing artwork. If that fails it will then ask the macOS
Quick Look system to generate a thumbnail to use as Now Playing artwork.
If that fails then it will try and use an OSC thumbnail.

This commit also corrects several existing problems with Now Playing
found during testing.

Co-authored-by: Matt Svoboda <matt.svoboda@gmail.com>

low-batt commented:
This commit is builds upon the pioneering work done by @svobs to add support for showing artwork in Now Playing. Obtaining artwork to show is a bit involved because the inputs needed for the process such as video tracks and OSC thumbnails appear asynchronously after the media is opened and work to search for and construct an artwork image is done in the background.

This commit definitely needs a critical review!

One difference between mpv and IINA is visible in the following screenshots…

mpv 0.4.0 supplies an image that is not square and Now Playing shows it with grey borders:
![mpv](https://github.com/user-attachments/assets/4fe81434-fc68-4af0-9bfe-6a5323bd7641)


IINA with this PR when possible crops the image to properly fit the space Now Playing provides:
![iina](https://github.com/user-attachments/assets/8c8993b5-acdb-46d4-a3f7-14b89d3afc82)


This has to do with the value that must be supplied for [MPMediaItemPropertyArtwork](https://developer.apple.com/documentation/mediaplayer/mpmediaitempropertyartwork), a [MPMediaItemArtwork](https://developer.apple.com/documentation/mediaplayer/mpmediaitemartwork) object. The [initializer](https://developer.apple.com/documentation/mediaplayer/mpmediaitemartwork/init(boundssize:requesthandler:)) takes a `boundsSize` parameter and a callback:
> **boundsSize**
>    The original size of the artwork.
>    
> **requestHandler**
> A handler that the system calls for the requested artwork.
>  &nbsp; &nbsp; &nbsp; &nbsp;**size**
>  &nbsp; &nbsp; &nbsp; &nbsp;The new size for the image.

Where `size` is a parameter to the `requestHandler` callback. Apple indicates:

> The request handler returns the image in the newly requested size. The requested size must be less than the `boundsSize` parameter.

Unfortunately this all the guidance I was able to find from Apple. 

mpv completely ignores the `size` parameter and returns the full sized image:

```swift
MPMediaItemArtwork(boundsSize: cover.size) { _ in return cover }
```

The implementation developed by @svobs crops and resizes the image to the `size` parameter. With verbose logging enabled you can see this:

```
17:27:06.051 [now-playing][v] Cropped and resized 2000x1125 artwork image to 768x768
```

In the above test case both IINA and mpv are displaying thumbnails generated by [QLThumbnailGenerator](https://developer.apple.com/documentation/quicklookthumbnailing/qlthumbnailgenerator). As a result the artwork matches up with the thumbnails shown in Finder:

![finder](https://github.com/user-attachments/assets/fabef718-94c5-4934-b9e4-85cf0e4d5f9f)

I have the Quick Look extension [QuickLook Video](https://github.com/Marginal/QLVideo) installed to gain thumbnail support for files not supported by Apple. It generated the thumbnails shown above.

If `QLThumbnailGenerator` fails to generate a thumbnail IINA will fall back to using an OSC thumbnail. The choice of playback position at which to obtain a thumbnail is similar to how `QuickLook Video` determines the position to use.

low-batt commented:
The mpv Now Playing implementation can be seen [here](https://github.com/mpv-player/mpv/blob/master/osdep/mac/remote_command_center.swift).

Note the loose coupling accomplished by using events. At some point IINA should consider making more use of events to break some of the tight coupling between IINA classes.

lhc70000 commented on iina/MPVTrack.swift:
```diff
@@ -121,11 +127,12 @@ class MPVTrack: NSObject {
   var decoderDesc: String?
 
-  init(id: Int, type: TrackType, isDefault: Bool, isForced: Bool, isSelected: Bool, isExternal: Bool) {
+  init(id: Int, type: TrackType, isDefault: Bool, isForced: Bool, isImage: Bool, isSelected: Bool, isExternal: Bool) {
```
I suggest adding a `convenience init` that matches the previous method signature and set `isImage` to false by default, so the above lines don't need to be changed.

low-batt replied:
Will do.

low-batt replied:
I added a default value of `false` for the new `isImage` parameter and removed the changes to the static constants.

uiryuu commented on iina/NowPlayingInfoManager.swift:
```diff
@@ -0,0 +1,620 @@
+      }
+    }
+
+    let duration = PlayerCore.lastActive.info.videoDuration?.second ?? 0
```
```suggestion
    let duration = activePlayer.info.videoDuration?.second ?? 0
```

low-batt replied:
For this PR I was trying to just add support for artwork and not cleanup the existing code. I ended cheating a bit on that and including some fixes because they were making it hard to test these changes. I'll go ahead and include these tweaks as well.

uiryuu commented on iina/NowPlayingInfoManager.swift:
```diff
@@ -0,0 +1,620 @@
+    info[MPNowPlayingInfoPropertyElapsedPlaybackTime] = time
+    info[MPNowPlayingInfoPropertyPlaybackRate] = speed
+    info[MPNowPlayingInfoPropertyDefaultPlaybackRate] = 1
+
```
What about adding `MPNowPlayingInfoPropertyAssetURL` here as well? Didn't notice any difference but the system might use this value.
```suggestion
    info[MPNowPlayingInfoPropertyAssetURL] = activePlayer.info.currentURL

```

uiryuu replied:
Also didn't notice any difference after adding `MPNowPlayingInfoPropertyChapterCount` and `MPNowPlayingInfoPropertyChapterNumber`, but those're also something we can add.

low-batt replied:
I'll add the setting of `MPNowPlayingInfoPropertyAssetURL` and then correct it to not be set if `currentURL` is `nil`.

There is a _lot_ that still needs to be done with respects to this code. This PR is focused on the artwork for the Now Playing module. Some of the missing settings like `MPNowPlayingInfoPropertyChapterCount` I don't think matter to the Now Playing module. _However_ I am expecting they do matter if you try and use Siri to control IINA. Originally I was going to open that can of worms for 1.4.0. I'm guessing there are only a few things that need fixing. But currently I was thinking of postponing looking into this as I still have several issues I need to address for beta2. Should I look into this?

uiryuu replied:
Let's focus on b2 for now.

uiryuu reviewed (approved):
Nice work. During my tests I didn't found any problem. See my comments and you can decide whether or not to add those info.

uiryuu commented:
Just notice a problem, sometimes when you open a new file, the previous artwork still remains. I don't know what triggers this, but it seems to appear more often when you start the new file paused.

low-batt committed:
Apply suggestions from code review

Co-authored-by: Yuze Jiang <i@uiryuu.moe>

low-batt commented:
On the wrong artwork being displayed, that is exactly what I've been worried about due to all the asynchronous activity. The `artworkTicket` is supposed to prevent that from happening. I tested a lot by setting IINA to pause at startup and dragging a lot of files onto the IINA icon, but so far have not be able to reproduce the problem with the wrong artwork being displayed. I specifically added a lot of verbose logging due to the potential for such a problem. I'll probably improve those log messages a bit. So far I only found one problem, I was not resetting the QL retries. I'll keep trying to reproduce this problem.

low-batt committed:
Close support cover artwork in Now Playing module, #5336

This commit will:
- Change updateInfo to ignore request if currentURL is not set
- Add missing code to reset artworkQLGenerationRetries
- Add support for setting MPNowPlayingInfoPropertyAssetURL,
  MPNowPlayingInfoPropertyPlaybackQueueCount,
  MPNowPlayingInfoPropertyPlaybackQueueIndex,
  MPNowPlayingInfoPropertyChapterCount and
  MPNowPlayingInfoPropertyChapterNumber
- Improve logging

low-batt commented:
I added support for more properties and fixed a problem with missing code to reset `artworkQLGenerationRetries`. So far I've not been able to reproduce a problem with the wrong artwork being displayed. I looked through the code and didn't spot a way in which that can happen. I improved the logging so that if it can be reproduced we might have a clue as to what is happening if verbose logging is enabled. I'll look over the code some more tomorrow.

One possibility is that this is a Now Playing glitch. I hit problems when a file is opened and IINA is configured to pause. Now Playing would ignore the update. That is why there is a workaround that tells Now Playing the media is playing and then immediately sets it to paused. As updating the artwork happens after that maybe Now Playing sometimes ignores that. There is code in mpv:
```swift
    func makeCurrent() {
        infoCenter.playbackState = .paused
        infoCenter.playbackState = .playing
        updateInfoCenter()
    }
```

What is that setting of `playbackState` all about? Some sort of work around?

low-batt commented:
@svobs Thoughts on this PR?

lhc70000 commented on iina/Extensions.swift:
```diff
@@ -813,3 +827,106 @@ extension Process {
+  }
+}
+
+extension MPNowPlayingPlaybackState: @retroactive CustomStringConvertible {
```
I think it's better to put this in NewPlayingInfoManager.swift

low-batt replied:
Will do.

low-batt replied:
This extension has been moved to `NowPlayingInfoManager`.

lhc70000 commented on iina/NowPlayingInfoManager.swift:
```diff
@@ -0,0 +1,620 @@
+      // NOTE this MUST be done before copying the nowPlayingInfo array below as abandonArtwork
+      // works directly on nowPlayingInfo.
+      if activePlayer.info.isNetworkResource {
+        abandonArtwork()
```
We have a setting `enableThumbnailForRemoteFiles` for thumbnail generation, maybe it should also be honored here.

low-batt replied:
I'll test, but I believe that is already supported as that is for mounted volumes, not network. I'll double check that.

I was going to task about `enableThumbnailForRemoteFiles` as there are several issues where this has come up. Is there a reason why we don't have UI for that setting?

low-batt replied:
The handling of `isNetworkResource` was incorrect. Streaming can have embedded artwork and there is a mpv option that allows you to specify artwork which will show up as a video track. The code now checks `isNetworkResource` and doesn't try and generate Quick Look thumbnails or search for OSC thumbnails when streaming, but will show any artwork found in tracks.

Where `enableThumbnailForRemoteFiles` comes into play is with mounted volumes. I could add code to check to see if the media is on a mounted volume and if so then not look for OSC thumbnails for it unless `enableThumbnailForRemoteFiles` is enabled. The main effect of that would only be changes in the log messages you see with verbose logging. Currently it just reports thumbnails are not ready, which is what happens unless `enableThumbnailForRemoteFiles` is enabled.

low-batt commented:
I am currently working on an update that fixes the problem @uiryuu found and cleans up some stuff. I'll also address the input from @lhc70000.

low-batt committed:
Close support cover artwork in Now Playing module, #5336

This commit will:
- Add a mpvStr property to the URL extension
- Move the MPNowPlayingPlaybackState from Extensions to
  NowPlayingInfoManager
- Add a default value of false for the new MPVTrack isImage parameter
- Remove passing of isImage in the MPVTrack static constants
- Rename the NowPlayingInfoManager property artworkURL to be url
- Rename the updateInfo constant activePlayer to just be player
- Remove checking of isNetworkResource before calling updateArtwork in
  updateInfo
- Rename abandonArtwork to discardArtwork to standardize naming
- Change updateInfo to detect when the media item changes
- Change artworkUpdateComplete to call updateInfo instead of
  updateArtwork
- Add resetting of artworkLastTrackCount to discardArtwork
- Change updateArtwork to not attempt to generate a Quick Look thumbnail
  or search for an OSC thumbnail if isNetworkResource is true
- Change PlayerCore.openMainWindow to reset many PlaybackInfo properties
- Updated and corrected comments and log messages

These changes correct multiple problems found during review.

low-batt committed:
Correct spelling errors.

low-batt commented:
The review comments gave me ideas for other tests to run. I found a number of problems. One of them was caused because `PlayerCore` did not reset `PlaybackInfo` properties when opening a new file.

low-batt committed:
Close support cover artwork in Now Playing module, #5336

This commit will change PlayerCore.openMainWindow to also set
videoHeight and videoWidth to nil.

svobs commented:
> One difference between mpv and IINA is visible in the following screenshots…
> 
> mpv 0.4.0 supplies an image that is not square and Now Playing shows it with grey borders
> 
> IINA with this PR when possible crops the image to properly fit the space Now Playing provides

In defense of cropping the image (in case anyone is thinking about this), I consider it the least bad option, given the constraints imposed by Apple. Drawing the image with gray space around it (or any other color) just ends up looking clunky and shoehorned in, IMHO. Apple clearly wants a square image to be shown there. And their API implies that they may change their mind in the future and apps should be prepared to deliver any size they ask for. So I think this is an improvement from what mpv does.

The code looks good! Only a couple of things:
1. The multithreading looks well done and should avoid tying up the main thread. But I did find that it took a while for the artwork to appear when I opened a large 4K HDR movie file. I think this happened because `artworkQueue` uses `qos: .background`, which is 1 level lower priority than `thumbnailQueue`, which uses `qos: .utility`. So Now Playing may have been left pending until the the thumbnails finishing generating. I'm thinking `artworkQueue` should at least be `utility` as well?
2. I played with a few test cases to try to get it to show stale artwork. Couldn't trigger it with any videos. But if I reused the window to play to an audio file which had no artwork, I found that Now Playing still showed the previous file's artwork. In my fork when there's no video track or there's a failure to generate an image, I just use the default album art for `MPMediaItemArtwork`. But I suppose another option might be to just clear it.

low-batt commented:
I'll change the queue priority.  The code is supposed to delete the artwork when the url changes. I'll look into why that did not happen.

low-batt commented:
I think everyone is good with the cripping. Note that it does not eliminate all use of those grey borders Now Playing seems to want 768x768. If the video is tiny and all that is available is the tiny OSC thumbnails, then both the video size and the thumbnail size are too small and you will end up with grey borders.

svobs commented:
> The code is supposed to delete the artwork when the url changes. I'll look into why that did not happen.

Looks like it properly clears the art if changing to the audio file via playlist navigation. I needed to drag & drop the audio file onto the video.

svobs commented:
> Note that it does not eliminate all use of those grey borders Now Playing seems to want 768x768. If the video is tiny and all that is available is the tiny OSC thumbnails, then both the video size and the thumbnail size are too small and you will end up with grey borders.

Really? Seemed fine for me with one of my nano-size test videos:

https://github.com/user-attachments/assets/f05cfd9b-6fbb-4454-beeb-986174b4a296

low-batt commented:
My guess is that this is what happened in your test run:

```
16:16:35.067 [now-playing][d] Found Quick Look thumbnail for: /Users/low-batt/Documents/builds/iina-official/issues/5336/450632367-f05cfd9b-6fbb-4454-beeb-986174b4a296.mp4
16:16:35.067 [now-playing][v] Video size (320x240) is smaller than image, using image size for bounds
```

IINA detected the image size was larger than the video size and used the image size for the bounds. That was a change I made. This worked because even through we are asking `QLThumbnailGenerator` for a `lowQualityThumbnail` it provides an image bigger than Now Playing needs:
```text
16:16:35.275 [now-playing][v] Cropped and resized 2000x1500 artwork image to 768x768
```

To see the problem you need a file that `QLThumbnailGenerator` does not support so IINA has to use an OSC thumbnail.  If I add a `, false` in `generateQLThumbnail` like so:
```swift
        guard let image = thumbnail?.nsImage, false else {
```

It will force use of an OSC thumbnail and:
```
16:25:27.435 [now-playing][v] Artwork bounds size: 320x240

16:25:27.637 [now-playing][v] Cropped and resized 240x180 artwork image to 320x240
```

I'm thinking we need to set the bounds size to have a minimum size of 768 and reflect the aspect ratio of the video or image in order to cause the image to be sized up.

Gotta run. Tuesday is my busy day.

low-batt commented:
@svobs I've not been able to reproduce the problem. I disabled `Switch to "music mode" automatically when playing an audio file`. I started playing a video. I dragged a music file onto the main window. The music file started playing. The artwork in Now Playing changed to the IINA icon.

There was such a problem with the initial version of the PR. Make sure you have the latest commits.

svobs commented:
> There was such a problem with the initial version of the PR. Make sure you have the latest commits.

My bad! Sorry. Updated to the latest version and confirmed that I am not seeing stale artwork now.

> To see the problem you need a file that QLThumbnailGenerator does not support so IINA has to use an OSC thumbnail. If I add a , false in generateQLThumbnail like so:

I see it now. Interesting. Seems that the `CGSize` parameter of the `MPMediaItemArtwork` request handler actually tries to preserve the aspect ratio from `boundsSize` for lower resolutions... That says something... Up to you if you want to try to improve on that design decision, but you could be forgiven for strictly adhering to the API and leaving it as-is.

low-batt committed:
Close support cover artwork in Now Playing module, #5336

This commit will:
- Change the qos for the artworkQueue to be utility instead of
  background to match the qos used for thumbnails
- Add an artworkDesiredSize constant representing the minimum image
  size desired by the Now Playing module
- Change formMPMediaItemArtwork to scale up the image size if needed
  when setting boundsSize

This insures boundsSize is large enough that the Now Playing module will
request the full size it supports and causing small OSC thumbnails
scaled up, avoiding Now Playing showing grey bars around the artwork.

low-batt commented:
The latest commit makes sure `boundsSize` is always large enough to avoid the grey bars. I also changed the queue priority to `utility`.

See if the new changes look good to go.
